### PR TITLE
Attempt to use the c version of qhull

### PR DESCRIPTION
**Pull request by GiulioRomualdi**
On ubuntu 22.04 I got the following error
```
usr/bin/ld: ../../../lib/libBipedalLocomotionFrameworkPlanners.so.0.14.200: undefined reference to `qh_new_qhull(int, int, double*, unsigned int, char*, _IO_FILE*, _IO_FILE*)'
/usr/bin/ld: ../../../lib/libBipedalLocomotionFrameworkPlanners.so.0.14.200: undefined reference to `qh_memfreeshort(int*, int*)'
/usr/bin/ld: ../../../lib/libBipedalLocomotionFrameworkPlanners.so.0.14.200: undefined reference to `qh_freeqhull(unsigned int)'
collect2: error: ld returned 1 exit status
make[3]: *** [src/Planners/tests/CMakeFiles/SwingFootPlannerUnitTests.dir/build.make:108: bin/SwingFootPlannerUnitTests] Error 1
make[2]: *** [CMakeFiles/Makefile2:3853: src/Planners/tests/CMakeFiles/SwingFootPlannerUnitTests.dir/all] Error 2
make[1]: *** [CMakeFiles/Makefile2:3860: src/Planners/tests/CMakeFiles/SwingFootPlannerUnitTests.dir/rule] Error 2
make: *** [Makefile:1141: SwingFootPlannerUnitTests] Error 2

```

with qhull installed via APT

### CODE REVIEW

**Comment by GiulioRomualdi** [comment]
I switched to qhull_r and the problem seems to be fixed. Now what is missing is a cmake function that allows the user to find qhull in ubuntu 20.04

**Review by S-Dafarra — APPROVED** [review]

**Comment by traversaro** [comment]
New failures:
~~~
[ 66%] Linking CXX executable ../../../bin/ConvexHullHelperUnitTests
/usr/bin/ld: ../../../lib/libBipedalLocomotionFrameworkPlanners.so.0.14.200: undefined reference to `qh_freeqhull(qhT*, unsigned int)'
/usr/bin/ld: ../../../lib/libBipedalLocomotionFrameworkPlanners.so.0.14.200: undefined reference to `qh_memfreeshort(qhT*, int*, int*)'
/usr/bin/ld: ../../../lib/libBipedalLocomotionFrameworkPlanners.so.0.14.200: undefined reference to `qh_zero(qhT*, _IO_FILE*)'
/usr/bin/ld: ../../../lib/libBipedalLocomotionFrameworkPlanners.so.0.14.200: undefined reference to `qh_new_qhull(qhT*, int, int, double*, unsigned int, char*, _IO_FILE*, _IO_FILE*)'
collect2: error: ld returned 1 exit status
~~~

**Comment by traversaro** [comment]
> New failures:
> 
> ```
> [ 66%] Linking CXX executable ../../../bin/ConvexHullHelperUnitTests
> /usr/bin/ld: ../../../lib/libBipedalLocomotionFrameworkPlanners.so.0.14.200: undefined reference to `qh_freeqhull(qhT*, unsigned int)'
> /usr/bin/ld: ../../../lib/libBipedalLocomotionFrameworkPlanners.so.0.14.200: undefined reference to `qh_memfreeshort(qhT*, int*, int*)'
> /usr/bin/ld: ../../../lib/libBipedalLocomotionFrameworkPlanners.so.0.14.200: undefined reference to `qh_zero(qhT*, _IO_FILE*)'
> /usr/bin/ld: ../../../lib/libBipedalLocomotionFrameworkPlanners.so.0.14.200: undefined reference to `qh_new_qhull(qhT*, int, int, double*, unsigned int, char*, _IO_FILE*, _IO_FILE*)'
> collect2: error: ld returned 1 exit status
> ```

As the symbols are in the library but just with C mangling and not with C++ one:
~~~
gitpod@traversaro-gitpodubuntu-j71zab8at4q:/usr/lib/x86_64-linux-gnu$ readelf -s libqhull_r.so.7 | grep qh_free
   229: 0000000000006880  1022 FUNC    GLOBAL DEFAULT   12 qh_freebuild
   357: 00000000000065f0   647 FUNC    GLOBAL DEFAULT   12 qh_freebuffers
   359: 000000000002f430     5 FUNC    GLOBAL DEFAULT   12 qh_free
   396: 0000000000006c80   138 FUNC    GLOBAL DEFAULT   12 qh_freeqhull
~~~

I think that we are missing `extern "C"` like in https://github.com/PointCloudLibrary/pcl/blob/3eabe5a772717028874db24a54bc6f88bffffae8/surface/include/pcl/surface/qhull.h#L49 . Probably this is not necessary in newer verson of qhull available in Ubuntu 22.04 as it is already included in the headers of qhull itself.

File: src/Planners/src/ConvexHullHelper.cpp [hunk 1]
```diff
@@ -5,11 +5,10 @@
  * distributed under the terms of the BSD-3-Clause license.
  */
 
-#include <libqhullcpp/PointCoordinates.h>
-#include <libqhullcpp/Qhull.h>
-#include <libqhullcpp/QhullFacetList.h>
+#include <libqhull_r/libqhull_r.h>
```
Comment (traversaro) [inline src/Planners/src/ConvexHullHelper.cpp:8]:
```suggestion
extern "C"
{
#include <libqhull_r/libqhull_r.h>
}
```

See https://github.com/ami-iit/bipedal-locomotion-framework/pull/713#issuecomment-1658386129 for the rationale.

File: .github/workflows/ci.yml [hunk 2]
```diff
@@ -128,8 +127,8 @@ jobs:
                              qtmultimedia5-dev libxml2-dev liburdfdom-dev libtinyxml-dev \
                              liburdfdom-dev liboctave-dev python3-dev valgrind coinor-libipopt-dev \
                              libmatio-dev python3-pytest python3-numpy python3-scipy \
-                             python3-setuptools  libspdlog-dev  libopencv-dev libpcl-dev \
-                             python3-pybind11 nlohmann-json3-dev libassimp-dev
+                             python3-setuptools  libspdlog-dev libopencv-dev libpcl-dev \
+                             python3-pybind11 nlohmann-json3-dev libassimp-dev libqhull-dev
```
Comment (traversaro) [inline .github/workflows/ci.yml:131]: This package only install the headers and `libqhull.so`, to have `libqhull_r.so` you also need to install `libqhull-r7` in old distro or `libqhull-r8.0` in newer one. However, probably this is already a dependency of pcl, so everything works.

File: src/Planners/src/ConvexHullHelper.cpp (hunk 1 above)
Comment (GiulioRomualdi) [reply src/Planners/src/ConvexHullHelper.cpp:10]: Thank you! Let's see if the CI is happy now

**Comment by traversaro** [comment]
Cool, now a runtime error:
~~~
2023-07-31T15:08:24.9658085Z QH6186 qhull error (qh_new_qhull): start qhull_cmd argument with "qhull "
~~~

File: src/Planners/src/ConvexHullHelper.cpp [hunk 3]
```diff
@@ -24,67 +26,106 @@ ConvexHullHelper::ConvexHullHelper()
...
-    std::vector<double> allCoordinates(numberOfPoints * numberOfCoordinates);
-    Eigen::Map<Eigen::MatrixXd>(allCoordinates.data(), numberOfCoordinates, numberOfPoints) = points;
+    // initialize qhull
+    qh_zero(qh, nullptr);
 
-    // map = points;
-    pointCoordinates.append(allCoordinates);
+    // set the input
+    char flags[250];
+    sprintf(flags, "qhull");
```
Comment (traversaro) [inline src/Planners/src/ConvexHullHelper.cpp:57]:
```suggestion
    sprintf(flags, "qhull ");
```

Hopefully this should fix the runtime behaviour with qhull from 20.04, see https://github.com/PointCloudLibrary/pcl/blob/3eabe5a772717028874db24a54bc6f88bffffae8/surface/include/pcl/surface/convex_hull.h#L131C44-L131C44 .

**Comment by GiulioRomualdi** [comment]
Thank you @traversaro I will squash and merge

**Pull request merged**